File: language/benchmarks/tokens/algorithms/histogram/histogram.py
Bucket = dict[str, int]


def buckets(histogram: dict[int, int]) -> list[Bucket]:
    return [{"count": value, "value": key} for key, value in histogram.items()]


def histogram(xs: list[int]) -> dict[int, int]:
    result: dict[int, int] = {}
    for value in xs:
        result = increment(result, value)
    return result


def increment(histogram: dict[int, int], value: int) -> dict[int, int]:
    next_histogram = dict(histogram)
    if value in next_histogram:
        next_histogram[value] = next_histogram[value] + 1
    else:
        next_histogram[value] = 1
    return next_histogram

File: language/benchmarks/tokens/algorithms/histogram/histogram.ts
type Bucket = { count: number; value: number };

function buckets(histogram: Map<number, number>): Bucket[] {
  return Array.from(histogram.entries()).map(([key, value]) => ({ count: value, value: key }));
}

function histogram(xs: number[]): Map<number, number> {
  return xs.reduce(increment, new Map<number, number>());
}

function increment(histogram: Map<number, number>, value: number): Map<number, number> {
  const nextHistogram = new Map(histogram);
  if (nextHistogram.has(value)) {
    nextHistogram.set(value, nextHistogram.get(value)! + 1);
  } else {
    nextHistogram.set(value, 1);
  }
  return nextHistogram;
}
